relay-cardinality/src/redis.rs

```diff
@@ -50,171 +49,256 @@ impl RedisSetLimiter {
+            drop(cache); // Give up the cache lock!
         }
 
         if !rejected.is_empty() {
             relay_log::debug!(
-                organization_id = organization_id,
+                scoping = ?scoping,
                 "rejected {} metrics due to cardinality limit",
                 rejected.len(),
             );
         }
 
-        acceptions_rejections.report(
-            CardinalityLimiterCounters::Accepted,
-            CardinalityLimiterCounters::Rejected,
-        );
-
         Ok(Box::new(rejected.into_iter()))
     }
 }
 
+/// A quota scoping extracted from a [`CardinalityLimit`] and a [`Scoping`].
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub(crate) struct QuotaScoping {
+    pub window: SlidingWindow,
+    pub namespace: Option<MetricNamespace>,
+    pub organization_id: Option<OrganizationId>,
+    pub project_id: Option<ProjectId>,
+}
+
+impl QuotaScoping {
+    /// Creates a new [`QuotaScoping`] from a [`Scoping`] and [`CardinalityLimit`].
+    ///
+    /// Returns `None` for limits with scope [`CardinalityScope::Unknown`].
+    pub fn new(scoping: Scoping, limit: &CardinalityLimit) -> Option<Self> {
+        let (organization_id, project_id) = match limit.scope {
+            CardinalityScope::Organization => (Some(scoping.organization_id), None),
+            // Invalid/unknown scope -> ignore the limit.
+            CardinalityScope::Unknown => return None,
+        };
+
+        Some(Self {
+            window: limit.window,
+            namespace: limit.namespace,
+            organization_id,
+            project_id,
+        })
+    }
+
+    /// Wether the scoping applies to the passed entry.
+    pub fn applies_to(&self, entry: &Entry) -> bool {
+        self.namespace.is_none() || self.namespace == Some(entry.namespace)
+    }
+
+    /// Returns all slots of the sliding window for a specific timestamp.
+    pub fn slots(&self, timestamp: u64) -> impl Iterator<Item = u64> {
+        self.window.iter(timestamp)
+    }
+
+    /// Turns the scoping into a Redis key for the passed slot.
+    fn to_redis_key(self, slot: u64) -> String {
+        let organization_id = self.organization_id.unwrap_or(0);
+        let project_id = self.project_id.map(|p| p.value()).unwrap_or(0);
+        let namespace = self.namespace.map(|ns| ns.as_str()).unwrap_or("");
+
+        format!("{KEY_PREFIX}:scope-{{{organization_id}-{project_id}-{namespace}}}-{slot}")
+    }
+}
+
+/// Internal state combining relevant entries for the respective quota.
+struct LimitState<'a> {
+    /// Id of the original limit.
+    pub id: &'a str,
+    /// Entries which are relevant for the quota.
+    pub entries: Vec<RedisEntry>,
+    /// Scoping of the quota.
+    pub scope: QuotaScoping,
+    /// The limit of the quota.
+    pub limit: u64,
+
+    /// Amount of cache hits `(hits, misses)`.
+    cache_hits: (i64, i64),
+    /// Amount of accepts and rejections `(acceptions, rejections)`.
+    acceptions_rejections: (i64, i64),
```

Comment:
Still skeptical about "acception", but apparently it is a word: https://en.wiktionary.org/wiki/acception.

Reply:
Ah sorry, forgot to adjust the names when changing the doc, will fix